woocommerce/woocommerce
# Add initial k6 performance tests

### All Submissions:






### Changes proposed in this Pull Request:



Closes https://github.com/woocommerce/woocommerce/issues/29170, https://github.com/woocommerce/woocommerce/issues/29492, https://github.com/woocommerce/woocommerce/issues/29999, https://github.com/woocommerce/woocommerce/issues/30000, https://github.com/woocommerce/woocommerce/issues/30164, https://github.com/woocommerce/woocommerce/issues/30240, https://github.com/woocommerce/woocommerce/issues/30238, https://github.com/woocommerce/woocommerce/issues/30429, and https://github.com/woocommerce/woocommerce/issues/30430.

More background: pcr5Cj-Fq-p2 & pcr5Cj-18n-p2

This PR contains an initial set of k6 performance tests for benchmarking performance (both single user and under load). 

Tests have been created in 'performance/requests' to simulate and measure the response time of the HTTP requests generated by the following steps:

Shopper
- Home Page (site root)
- Shop Page 
- Product Search
- Individual Product Page
- Add to Cart
- Remove from Cart
- Apply Coupon to Cart
- View Cart
- Proceed to Checkout
- Place Order (as guest)
- Login during Checkout
- Place Order (as existing customer)
- Order Received Page
- My Account (as existing customer)

Merchant
- WP Login (redirecting to WP Admin)
- WooCommerce Home
- Products Page
- Add New Product
- Orders Page

This means the following requests are used

Shopper
GET `/`
GET `/shop`
GET `/?s(product name)&post_type=product`
GET `/product/(product name)`
POST `/?wc-ajax=add_to_cart`
GET `/cart`
POST `/cart`
GET `/cart/?removed_item=`
POST `/?wc-ajax=apply_coupon`
GET `/checkout`
POST `/checkout`
POST `/?wc-ajax=update_order_review`
POST `/?wc-ajax=checkout`
GET `/checkout/order-received`
POST `/?wc-ajax=get_refreshed_fragments`
GET `/my-account`
POST `/my-account`

Merchant
GET `/wp-login.php`
POST `/wp-login.php`
GET `/wp-admin/admin.php?page=wc-admin`
GET `/wp-json/wc-analytics/admin/notes?`
GET `/wp-json/wc-analytics/orders?`
GET `/wp-json/wc-analytics/products/reviews?`
GET `/wp-json/wc-analytics/products?`
GET `/wp-json/wc-admin/options?`
POST `/wp-admin/admin-ajax.php`
GET `/wp-admin/edit.php?post_type=shop_order`
GET `/wp-admin/edit.php?post_type=product`
GET `/wp-admin/post-new.php?post_type=product`
POST `/wp-admin/post.php`

### How to test the changes in this Pull Request:

Before you try using the tests make sure you have an environment to run against.

In this PR I have included a Docker Initialization Script `init-sample-products.sh` that will set up a store with sample products imported and the store settings (payment method, permalinks, address etc) we need already set. I recommend using this if you just want to see the tests in action.
     
`npx wc-e2e docker:up ./tests/e2e/docker/init-sample-products.sh`

If you use a different environment you can change the details in `config.js`. 

For the purpose of these initial tests which follow on from the proof of concept you do not need k6 installed as we are going to use the k6 docker image to execute the test.

`docker pull loadimpact/k6`

To run the Shopper tests
`docker run --network="host"  -v /[YOUR LOCAL WC DIRECTORY FULL PATH]/tests:/tests -it loadimpact/k6 run /tests/performance/tests/simple-all-shopper-requests.js`

To run the Merchant tests
`docker run --network="host"  -v /[YOUR LOCAL WC DIRECTORY FULL PATH]/tests:/tests -it loadimpact/k6 run /tests/performance/tests/simple-all-merchant-requests.js`

### Other information:




### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.
